opal/rocm/common_rocm.c

```diff
@@ -0,0 +1,318 @@
+
+    ret = opal_dl_lookup (hip_handle, "hipFree", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipFree\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipFree) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipMemcpy", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipMemcpy\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipMemcpy) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipMemcpyAsync", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipMemcpyAsync\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipMemcpyAsync) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipStreamCreate", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipStreamCreate\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipStreamCreate) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipStreamDestroy", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipStreamDestroy\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipStreamDestroy) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipStreamSynchronize", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipStreamSynchronize\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipStreamSynchronize) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipGetErrorString", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipGetErrorString\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipGetErrorString) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipPointerGetAttributes", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(10, "Failed to find hipPointerGetAttributes\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipPointerGetAttributes) = ptr;
+
+    return OPAL_SUCCESS;
+}
+
+static void hip_dl_finalize(void)
+{
+    hipError_t err = hip_funcs.hipStreamDestroy(&common_rocm_MemcpyStream);
+    if (hipSuccess != err) {
+        opal_output(10, "hip_dl_finalize: error while destroying the hipStream\n");
+    }
+    dlclose(hip_handle);
+}
+
+
+static void mca_common_rocm_support_init(void)
+{
+    if (rocm_initialized) {
+        return;
+    }
+
+    OBJ_CONSTRUCT(&common_rocm_init_lock, opal_mutex_t);
+    OPAL_THREAD_LOCK(&common_rocm_init_lock);
+
+    if (rocm_initialized) {
+        return;
+    }
+
+    if (!OPAL_HAVE_DL_SUPPORT) {
+        opal_output(1, "Can not open libamdhip64.so without dlopen support\n");
```

Comment:
1? The code below uses 10.

Reply:
but either one is a bug; opal_output()'s first argument should almost never be an integer constant.  With the possible exception of 0/-1.

Reply:
fixed, thanks for catching this!

Reply:
> but either one is a bug; opal_output()'s first argument should almost never be an integer constant. With the possible exception of 0/-1.

ok, let me revisit this aspect

Reply:
I changed the opal_output streams to 0